api/client.go

```diff
@@ -356,29 +365,52 @@ func (c *Client) newRequest(
 }
 
 func (c *Client) doRequest(req *http.Request) (*http.Response, error) {
-	resp, err := c.httpClient.Do(req)
-	if err != nil {
-		return nil, err
-	}
+	operation := func() (*http.Response, error) {
+		resp, err := c.httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
 
-	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-		body, _ := io.ReadAll(resp.Body)
-		resp.Body.Close()
+		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+			body, _ := io.ReadAll(resp.Body)
+			resp.Body.Close()
 
-		if isAuthenticationRequiredResponse(resp) {
-			return nil, ErrAuthenticationRequired
-		}
-		if isGatewayTimeoutResponse(resp) {
-			return nil, ErrGatewayTimeout
-		}
-		if isNotFoundResponse(resp) {
-			return nil, ErrNotFound
+			switch resp.StatusCode {
+			case http.StatusUnauthorized:
+				return nil, backoff.Permanent(ErrAuthenticationRequired)
+
+			case http.StatusNotFound:
+				return nil, backoff.Permanent(ErrNotFound)
+
+			case http.StatusGatewayTimeout:
+				return nil, ErrGatewayTimeout
+
+			case http.StatusTooManyRequests:
+				if reset, e := strconv.ParseInt(resp.Header.Get("X-Ratelimit-Reset"), 10, 0); e == nil && reset > 0 {
+					return nil, backoff.RetryAfter(int(time.Until(time.Unix(int64(reset), 0)).Seconds()))
+				}
+
+				return nil, ErrRateLimitExceeded
+			}
+
+			return nil, backoff.Permanent(fmt.Errorf("request failed with status %d: %s", resp.StatusCode, body))
 		}
 
-		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, body)
+		return resp, nil
+	}
+
+	maxRetries := uint(5)
+	if req.Body != nil && req.GetBody == nil { // Body cannot be reset, so we can only read it once
+		maxRetries = 1
 	}
 
-	return resp, nil
+	return backoff.Retry(
+		req.Context(),
+		operation,
+		backoff.WithMaxTries(maxRetries),
+		backoff.WithMaxElapsedTime(10*time.Second),
```

Comment:
This might be a little too little if the `X-Ratelimit-Reset` header is set to next minute.